onnxruntime/core/graph/graph.cc

```diff
@@ -1893,11 +1893,29 @@ Status Graph::InferAndVerifyTypeMatch(Node& node, const OpSchema& op, const Reso
         } else {
           output_def->SetType(inferred_type);
         }
-      } else
-        return Status(ONNXRUNTIME, FAIL,
-                      "Type Error: Type (" + *existing_type + ") of output arg (" +
-                          output_def->Name() + ") of node (" + node_name +
-                          ") does not match expected type (" + *inferred_type + ").");
+      }
+
+      else {
+        // Make an exception for cases where the inferred type is `float16` and the existing
+        // type is `float` and the node is assigned to the CPU EP.
+        // We allow such cases to pass through without a complaint because there are some ONNX ops
+        // - mostly Generator ops (RandomNormal, RandomNormalLike, EyeLike, etc.) where-in
+        // the output type is inferred based on a static attribute (for example - `dtype`) as opposed
+        // to most other ONNX ops where the-in the type is deduced based on input types.
+        // When we insert cast nodes to the inputs of such operators (when the input type
+        // is float16) to cast the inputs to float so as to use CPU kernels (when there is no EP capable
+        // of running such nodes as is), the output NodeArg(s) are of the type `float` (existing_type) but the
+        // inferred type will be `float16` if such nodes have the output type defining attribute to be `float16`.
+        // We make an exception for this case and use the existing_type.
+        if (*inferred_type != "tensor(float16)" ||
+            *existing_type != "tensor(float)" ||
+            node.GetExecutionProviderType() != kCpuExecutionProvider) {
```

Comment:
Would it be better to instead fix the mismatch instead of ignoring it? 

- Could check node has one output edge and that is to a Cast node that converts the float back to float16  
  - not sure if the Cast to undo the type change is always the very next node 
- Or check there was a Cast node immediately prior to this node that converted from float16 to float
  - not sure if that Cast could be further back 
- Or maybe when we insert the Cast that transformer needs to special case these ops and update the relevant node attribute. 
  - More work, but would keep the graph correct and 'fixes' things at the point they're made invalid

Reply:
Synced offline. Went with the third approach.